### ISSUE TITLE
audio doesn't work with game-activity

### ISSUE THREAD

**Issue by tshrpl** [rib/android-activity#7]
i tried to get cpal working but couldn't do so

edit:
code: https://github.com/tshrpl/duel (run the game crate)

**Comment by rib** [comment]
Just to check - is cpal working ok with this app when run as a desktop app?

**Comment by tshrpl** [comment]
yes

**Comment by rib** [comment]
okey, yeah, curious. I haven't played with opal/cpal on Android but have been using cpal in a NES emulator recently so would be curious to try.

**Comment by tshrpl** [comment]
please, my game would never be complete without audio

**Comment by rib** [comment]
Okey, I've just pushed an Oboe based example that demonstrates playing a 440Hz sine wave on Android, running with GameActivity: https://github.com/rib/android-activity/tree/main/examples/agdk-oboe

Testing that on my phone it seems to work.

Since cpal uses Oboe then I would hope that cpal should also work but I've only tried using Opal directly so far.

It would be interesting to hear if the above example works for you.

**Comment by rib** [comment]
not sure but maybe the next step should be to try modifying the minimal oboe example to use cpal, based on the cpal android example: https://github.com/RustAudio/cpal/blob/master/examples/android.rs and if that works maybe experiment with playing back your wav file instead of playing a generated sine wave.

**Comment by tshrpl** [comment]
i copied the cpal rust example directly into `audio.rs`, just renamed the `main` function to `try_start` and inserted `std::thread::spawn(|| audio::try_start());` in `MainEvent::Resume` in `lib.rs`. it builds and the logs dont show any signigicant error expect this
```rs
INFO [android_debugger] RustStdoutStderr: thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', C:\Users\hitus\.cargo\registry\src\github.com-1ecc6299db9ec823\ndk-glue-0.6
.2\src\lib.rs:60:39
```
i dont here any sound

**Comment by rib** [comment]
I just had a chance to give this a try too and for some reason it worked ok for me: https://github.com/rib/android-activity/pull/11

There was one recent cpal change to bump it up to ndk 0.7 that I saw but maybe there were also some other changes since last trying.

@tshrpl I wonder if you could confirm if the cpal example in the above branch works for you?

**Comment by rib** [comment]
oh, no, my bad I made a mistake with my testing and I actually get the same error that you got @tshrpl

**Comment by rib** [comment]
I also only just noticed that the panic is coming from `ndk-glue`! which is pretty weird here. As far as I could tell ndk-glue is only a dev dependency in cpal, which is just for their android example as far as I can tell.

This almost seems like a cargo bug to me, since dev-dependencies aren't supposed to be inherited by other packages.

If I explicitly remove the ndk-glue dev dependency from cpal then I do actually get some audio output though after about 1 second theres a different crash I haven't looked at yet.

I'm a bit unsure about the right way to deal with the first issue though since that seems to be caused by Cargo. Quoting from the Cargo reference manual it says:

```
Dev-dependencies are not used when compiling a package for building, but are used for compiling tests, examples, and benchmarks.

These dependencies are not propagated to other packages which depend on this package.
```

This crash seems to be caused by a violation of that last point because the cdylib we're building in one package is pulling in ndk-glue via a dev-dependency of another package.

**Comment by rib** [comment]
For the second crash I see - I'm guessing at the moment that there is something unsafe/unsound with the cpal bindings, since I can certainly imagine various ways this crude example code might trigger a use-after-free. I suppose the cpal bindings over oboe might not be that safe.

I'm guessing it'll be much more likely to work by moving the device up into the event loop/main scope instead of trying to synchronously create a temporary device and play audio for one second and hope that it all cleans up neatly when dropped.

I'm guessing there's asynchronous work oboe is still trying to do to play audio samples while all the cpal state is being dropped which leads to the crash.

**Comment by rib** [comment]
To help workaround this issue, I've created a branch of cpal that just removes the ndk-glue dev dependency: https://github.com/rib/cpal/tree/no-ndk-glue-dev-dependency

**Comment by rib** [comment]
For reference, I've updated the cpal example in the PR: https://github.com/rib/android-activity/pull/11 so it now uses the above workaround branch of cpal and the code has been re-worked to find a device and create a stream up-front.

I'm finding that it's still crashy when suspending and resuming the app but it's better than it was before :) I think maybe the cpal bindings might just be a bit buggy/unsafe at the moment - maybe since they aren't heavily tested on Android currently.

**Comment by rib** [comment]
For reference here, I've filed a Cargo issue since it really seems to be going against the documentation for how dev-dependencies are intended to work: https://github.com/rust-lang/cargo/issues/10980

**Comment by rib** [comment]
Incidentally; if I go back to check the minimal `agdk-oboe` example, that also crashes when pausing the app, and from seeing this issue too: https://github.com/katyo/oboe-rs/issues/41 I wonder now whether the crash here (not the one related to ndk-glue) is going to turn out to be an oboe issue not a cpal issue.

**Comment by rib** [comment]
oh, actually this issue is sounding more relevant: https://github.com/katyo/oboe-rs/issues/32

**Comment by rib** [comment]
okey, so it turned out second crash I was chasing (not the ndk-glue one) was a bit of a red-herring. There was a more general bug causing a crash in _all_ the examples when suspending + resuming, which was caused by an `ndk 0.7` change that updated NativeWindow to implement Drop, which now unconditionally calls _release on the underlying ANativeWindow when the wrapper type is dropped. I've just pushed a fix for this to main: https://github.com/rib/android-activity/commit/93828a18a9be395d7bd9bb660f4926f38b5fe85a

With that fix then I see that the cpal example seems to work ok now (still with the workaround for the Cargo/ndk-glue issue) so I'll probably merge the PR for the cpal example now.

**Comment by rib** [comment]
Okey, although we still have this awkward situation where we need a workaround in cpal to avoid the `ndk-glue` dev dependency we now have cpal and oboe examples that seem to at least be able to play a simple tone which hopefully at least validates that they are usable with `android-activity`.

ref: https://github.com/rib/android-activity/pull/11

**Comment by rib** [comment]
As a quick follow up here - it turned out I was jumping the gun with thinking there was an issue with cpal leaking its dev dependency - actually the 0.13.5 version of cpal has ndk-glue as a regular dependency, not just a dev-dependency so when I commented out the dev-dependency and found it worked that was actually redundant and it also works fine to just use cpal master instead of the released 0.13.5 version.

Unfortunately they have done a release yet since that change so the cpal example will still have to fetch cpal from git for now.
